### PR TITLE
[ROCm] Improve softmax performance

### PR DESCRIPTION
**Pull request by doru1004**
This patch improves the performance of softmax for 2D tensors by:
- using a softmax calculation which eliminates the increase of shared memory usage with the size of the tensor and relies on global memory accesses for the tensor data accesses while still using shared memory for the actual reduction step (the shared memory used for the reduction is constant and does not increase with tensor size).
- for the final computation replacing the division by the sum with the multiplication of 1/sum. The 1/sum is computed as the last step of the warp reduction.
- replace the use of the exp function with the __expf function.

The impact on numerical accuracy is within a 1e-5 for half precision and 1e-7 for full precision.

The impact on performance for MI300X is between 22% and 50% percentage improvement over current runtimes.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd

### CODE REVIEW

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 1]
```diff
@@ -858,23 +959,30 @@ Tensor host_softmax(const Tensor & input_, const int64_t dim_, const bool half_t
...
-            can_use_smem &= (!(reinterpret_cast<uintptr_t>(output_ptr) % ALIGN_BYTES));
-            can_use_smem &= !(dim_size % ILP);
-
-            if (can_use_smem) {
-              size_t smem_sz = dim_size * sizeof(scalar_t) + smem_reduction_sz;
-              cunn_SoftMaxForwardSmem<ILP, scalar_t, accscalar_t, scalar_t, Epilogue>
-                <<<grid, block, smem_sz, stream>>>(output_ptr, input_ptr, dim_size);
-            } else {
-              cunn_SoftMaxForward<ILP, scalar_t, accscalar_t, scalar_t, Epilogue>
+            if (use_fast_softmax) {
```
Comment (jeffdaily) [inline aten/src/ATen/native/cuda/SoftMax.cu:962]: Would the compiler treat this the same way as if constexpr?  The `use_fast_softmax` is a template param so constexpr seems appropriate here and elsewhere it is used like this.
```suggestion
            if constexpr (use_fast_softmax) {
```

**Comment by jeffdaily** [comment]
CI will build these changes but not test them unless the build is temporarily changed to define PYTORCH_USE_FAST_SOFTMAX.

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 2]
```diff
@@ -894,23 +1002,30 @@ Tensor host_softmax(const Tensor & input_, const int64_t dim_, const bool half_t
...
-            can_use_smem &= (!(reinterpret_cast<uintptr_t>(output_ptr) % ALIGN_BYTES));
-            can_use_smem &= !(dim_size % ILP);
-
-            if (can_use_smem) {
-              size_t smem_sz = dim_size * sizeof(scalar_t) + smem_reduction_sz;
-              cunn_SoftMaxForwardSmem<ILP, scalar_t, accscalar_t, accscalar_t, Epilogue>
-                <<<grid, block, smem_sz, stream>>>(output_ptr, input_ptr, dim_size);
-            } else {
-              cunn_SoftMaxForward<ILP, scalar_t, accscalar_t, accscalar_t, Epilogue>
+            if (use_fast_softmax) {
```
Comment (jeffdaily) [inline aten/src/ATen/native/cuda/SoftMax.cu:1005]:
```suggestion
            if constexpr (use_fast_softmax) {
```
another place where if constexpr could be used similar to the above change

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 3]
```diff
@@ -1093,7 +1208,11 @@ TORCH_IMPL_FUNC(softmax_cuda_out) (
   const int64_t dim,
   const bool half_to_float,
   const Tensor &output) {
-  host_softmax<SoftMaxForwardEpilogue,false>(input, dim, half_to_float, output);
+#if defined(USE_ROCM) && defined(PYTORCH_USE_FAST_SOFTMAX)
```
Comment (jeffdaily) [inline aten/src/ATen/native/cuda/SoftMax.cu:1211]: can we temporarily make this just USE_ROCM so the CI will run with these changes?  putting it behind the PYTORCH_USE_FAST_SOFTMAX means none of this new code will be exercised.

Comment (doru1004) [reply aten/src/ATen/native/cuda/SoftMax.cu]: Just removed that temporarily.

**Comment by doru1004** [comment]
New PR: https://github.com/pytorch/pytorch/pull/149076
